bchrpc/server.go

```diff
@@ -1113,16 +1208,421 @@ func (s *GrpcServer) GetMerkleProof(ctx context.Context, req *pb.GetMerkleProofR
+
+		if slpMsg.TransactionType == "SEND" {
+			if len(slpMsg.Data.(v1parser.SlpSend).Amounts) < int(query.PrevOutVout) {
+				return nil, status.Error(codes.Aborted, "vout is not a valid SLP output")
+			}
+			result.TokenId = slpMsg.Data.(v1parser.SlpSend).TokenID
+			result.ValidityResultType = &pb.GetTrustedSlpValidationResponse_ValidityResult_V1TokenAmount{
+				V1TokenAmount: slpMsg.Data.(v1parser.SlpSend).Amounts[query.PrevOutVout-1],
+			}
+		} else if slpMsg.TransactionType == "MINT" {
+			result.TokenId = slpMsg.Data.(v1parser.SlpMint).TokenID
+			if query.PrevOutVout == 1 {
+				result.ValidityResultType = &pb.GetTrustedSlpValidationResponse_ValidityResult_V1TokenAmount{
+					V1TokenAmount: slpMsg.Data.(v1parser.SlpMint).Qty,
+				}
+			} else if int(query.PrevOutVout) == slpMsg.Data.(v1parser.SlpMint).MintBatonVout {
+				result.ValidityResultType = &pb.GetTrustedSlpValidationResponse_ValidityResult_V1MintBaton{
+					V1MintBaton: true,
+				}
+			} else {
+				return nil, status.Error(codes.Aborted, "vout is not a valid SLP output")
+			}
+		} else if slpMsg.TransactionType == "GENESIS" {
+			hash := query.PrevOutHash
+			for i := len(hash) - 1; len(result.TokenId) < len(hash); i-- {
+				result.TokenId = append(result.TokenId, hash[i])
+			}
+			if query.PrevOutVout == 1 {
+				result.ValidityResultType = &pb.GetTrustedSlpValidationResponse_ValidityResult_V1TokenAmount{
+					V1TokenAmount: slpMsg.Data.(v1parser.SlpGenesis).Qty,
+				}
+			} else if int(query.PrevOutVout) == slpMsg.Data.(v1parser.SlpGenesis).MintBatonVout {
+				result.ValidityResultType = &pb.GetTrustedSlpValidationResponse_ValidityResult_V1MintBaton{
+					V1MintBaton: true,
+				}
+			} else {
+				return nil, status.Error(codes.Aborted, "vout is not a valid SLP output")
+			}
+		}
+		result.SlpTxnOpreturn = entry.SlpOpReturn
+		results[i] = result
+	}
+	resp.Results = results
+	return resp, nil
+}
+
+// GetBip44HdAddress this method will return an address based on the requested HD path
+func (s *GrpcServer) GetBip44HdAddress(ctx context.Context, req *pb.GetBip44HdAddressRequest) (*pb.GetBip44HdAddressResponse, error) {
+	xpub := req.Xpub
+	if len(xpub) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "xpub is missing in request")
+	}
+
+	if xpub[:4] != "xpub" {
+		return nil, status.Error(codes.InvalidArgument, "xpub provided does not start with 'xpub'")
+	}
+
+	masterKey, err := hdkeychain.NewKeyFromString(xpub)
+	if err != nil {
+		fmt.Println(err)
+		return nil, status.Error(codes.InvalidArgument, "invalid xpub: "+err.Error())
+	}
+
+	var change uint32 = 0
+	if req.Change {
+		change = 1
+	}
+
+	ext, err := masterKey.Child(change)
+	if err != nil {
+		fmt.Println(err)
+		return nil, status.Error(codes.InvalidArgument, "invalid xpub: "+err.Error())
+	}
+
+	extK, err := ext.Child(req.AddressIndex)
+	if err != nil {
+		fmt.Println(err)
+		return nil, status.Error(codes.InvalidArgument, "invalid xpub: "+err.Error())
+	}
+
+	pubKey, _ := extK.ECPubKey()
```

Comment:
Errors! We just need to check all the spots. Because Go doesn't have exceptions, some of the errors that could be raised might surprise you!

Reply:
done.